src/core/geometry/qgsregularpolygon.h

```diff
@@ -0,0 +1,213 @@
+ *   the Free Software Foundation; either version 2 of the License, or     *
+ *   (at your option) any later version.                                   *
+ *                                                                         *
+ ***************************************************************************/
+
+#ifndef QGSREGULARPOLYGON_H
+#define QGSREGULARPOLYGON_H
+
+#include <QString>
+
+#include "qgis_core.h"
+#include "qgspointv2.h"
+#include "qgspolygon.h"
+#include "qgslinestring.h"
+#include "qgscircle.h"
+#include "qgstriangle.h"
+
+enum ConstructionOption
+{
+  inscribedCircle,
+  circumscribedCircle
+};
+
+/** \ingroup core
+ * \class QgsRegularPolygon
+ * \brief Regular Polygon geometry type.
+ *
+ * A regular polygon is defined by a center point with a number of sides/vertices, a radius and the first vertice.
+ *
+ * \since QGIS 3.0
+ */
+class CORE_EXPORT QgsRegularPolygon
+{
+  public:
+    QgsRegularPolygon();
+
+    /** Constructs a regular polygon by center and parameters for the first vertice. An empty regular polygon is returned if numSides < 3 or circle option isn't valid.
+     * \param center The center of the regular polygon.
+     * \param radius Distance from the center and the first vertice or sides (see \a circle).
+     * \param azimuth Angle in degrees started from the North to the first vertice.
+     * \param numSides Number of sides of the regular polygon.
+     * \param circle Option to create the polygon: inscribed in circle (the radius is the distance between the center and vertices) or circumscribed about circle (the radius is the distance from the center to the midpoints of the sides).
+     */
+    QgsRegularPolygon( const QgsPointV2 center, const double radius, const double azimuth, const int numSides, const int circle );
+
+    /** Constructs a regular polygon by center and another point.
+     * \param center The center of the regular polygon.
+     * \param pt1 The first vertice if the polygon is inscribed in circle or the midpoint of a side if the polygon is circumscribed about circle.
+     * \param numSides Number of sides of the regular polygon.
+     * \param circle Option to create the polygon inscribed in circle (the radius is the distance between the center and vertices) or circumscribed about circle (the radius is the distance from the center to the midpoints of the sides).
+     */
+    QgsRegularPolygon( const QgsPointV2 center, const QgsPointV2 pt1, const int  numSides, const int circle );
+
+    /** Constructs a regular polygon by two points of the first side.
+     * \param pt1 The first vertice of the first side, also first vertice of the regular polygon.
+     * \param pt2 The second vertice of the first side.
+     * \param numSides Number of sides of the regular polygon.
+     */
+    QgsRegularPolygon( const QgsPointV2 pt1, const QgsPointV2 pt2, const int  numSides );
+
+    virtual bool operator ==( const QgsRegularPolygon &rp ) const;
```

Comment:
Do we need these virtual methods? As far as I can see this class isn't going to be subclassed, so removing the virtual keyword will optimise the class